[글로] 3, 4단계 영화 극장 선택 제출합니다.

ippnsj opened the pull request:
안녕하세요 리뷰어님~! 😄
이번 미션에서 MVP도, Data Binding 도 처음이다 보니 많이 헤맸던 것 같아요..🥹
제가 어려워한 점을 정리해보면 다음과 같습니다!

1. 어디까지를 Presenter의 역할로 두어야 할 지가 모호했습니다!
사용자의 입력으로 Model 이 바뀌지 않는 순간에도 Presenter 를 거쳐야 하는건지 아닌지 잘 모르겠더라구요ㅠㅠ
Presenter 를 거치게 되면 Presenter 가 너무 비대해지고 액티비티가 하는 일이 없어진다고 생각하여 굳이 Presenter 를 거칠 필요가 없다고 판단하였습니다..! 🙃
2. Data Binding 을 처음 사용해보다 보니 장점을 살리지 못하고 View Binding 처럼 사용한 것 같아요ㅠㅠ
굳이 양방향일 필요가 없다고 판단한 곳에서는 View Binding 처럼 구현하게 되었습니다...ㅎㅎ  🙃

제 판단이 잘못되었다면 알려주세요! 리뷰어님의 도움이 필요합니다! 🤓

ippnsj committed:
refactor: 영화 상영 알람 시간 구하는 로직 분리

ippnsj committed:
refactor: 알람 권한 요청 함수 분리

ippnsj committed:
refactor: Notification Manager 클래스 분리

ippnsj committed:
refactor: SettingsFragment 에서 SettingsStorage 종속성 제거

ippnsj committed:
refactor: MovieDetailActivity 에서 PeopleCount 종속성 제거

ippnsj committed:
refactor: MovieDetailActivity 에서 Times 종속성 제거

ippnsj committed:
fix: 알람 허용 설정 Intent 시작 버그 수정

권한 설정창 대신 알람 허용 설정창 띄워줌
최소 버전에 맞게 패키지명 전달하는 코드 수정

ippnsj committed:
refactor: 두 날짜 사이 날짜를 모두 가져오는 로직 domain 에서 ui 로 이동

ippnsj committed:
refactor: SeatPickerActivity 에서 MovieTicket 종속성 제거

ippnsj committed:
fix: 조조 및 야간 할인 정책 티켓마다 적용

ippnsj committed:
refactor: 선택된 좌석 뷰 설정 함수로 분리

ippnsj committed:
fix: 평일과 주말 시간 리스트 수정

평일에는 10시부터 22시까지 2시간 간격
주말에는 9시부터 23시까지 2시간 간격

ippnsj committed:
test: MovieTicket 테스트 수정 및 삭제

ippnsj committed:
refactor: fragment 를 add 와 show 를 같이 하지 않도록 수정

ippnsj committed:
docs: step4 기능 목록 작성

ippnsj committed:
feat: 영화 선택 레이아웃 작성

ippnsj committed:
feat: 극장 선택 기능 구현

ippnsj committed:
refactor: MovieDetailActivity 에 DataBinding 적용

ippnsj committed:
refactor: seat selector 구현

ippnsj committed:
refactor: seat picker done selector 구현

ippnsj committed:
refactor: SeatPickerActivity 에 DataBinding 적용

ippnsj committed:
refactor: 극장 선택 프래그먼트 생성 캡슐화

ippnsj committed:
refactor: 영화와 극장 아이템 DataBinding 적용

ippnsj committed:
refactor: SeatPickerActivity 에서 Reservations 종속성 제거

ippnsj committed:
feat: 기존 예매 내역 유지 기능 구현

vagabond95 commented on app/src/main/res/layout/activity_seat_picker.xml:
```diff
@@ -1,316 +1,319 @@
+                    tools:layout_conversion_absoluteHeight="34dp"
+                    tools:layout_conversion_absoluteWidth="292dp" />
+
+                <TableLayout
```
table row 를 코드 레벨에서 생성하도록 리펙토링해보는 것은 어떨까요?

모든 seat 에 대한 view 를 xml 레벨에서 관리할 경우 현재 처럼 많은 diff 가 발생하고
요구사항을 반영하기에도 불편할 것 같아요.

vagabond95 commented on app/src/main/res/layout/activity_movie_ticket.xml:
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
해당 xml 에는 <layout> 태그로 감싸주지 않으신 이유가 있을까요?

vagabond95 commented on app/src/main/java/woowacourse/movie/MovieApplication.kt:
```diff
@@ -1,11 +1,18 @@
+
         SettingsStorage.init(this)
+
+        val db = DBHelper(this).readableDatabase
+        Reservations.restore(db)
+        db.close()
```
Application 레벨의 코드는 앱이 실행될 때마다, 호출이 되기 때문에 되도록 어떠한 처리도 진행되지 않는 것이 좋습니다.

vagabond95 replied:
특히나 db 같이 처리속도가 무겁고 느린 코드의 경우, 더더욱 지양하셔야될 코드라고 생각됩니다.

vagabond95 commented on app/src/main/java/woowacourse/movie/ui/fragment/settings/presentation/SettingsPresentation.kt:
```diff
@@ -0,0 +1,12 @@
+import woowacourse.movie.data.storage.SettingsStorage
+import woowacourse.movie.ui.fragment.settings.SettingsContract
+
+class SettingsPresentation : SettingsContract.Presentation {
```
mvp 패턴이라면, presenter 가 view 에게 명령을 내릴 수 있어야하는데
현재 presenter 는 view 에 대한 참조가 없는 상황입니다.

vagabond95 commented on app/src/main/java/woowacourse/movie/ui/fragment/settings/view/SettingsFragment.kt:
```diff
@@ -22,14 +24,16 @@ class SettingsFragment : Fragment() {
     private fun initPushNotificationSwitch(view: View) {
         val pushNotificationSwitch = view.findViewById<SwitchCompat>(R.id.push_notification_switch)
-        pushNotificationSwitch.isChecked = SettingsStorage.enablePushNotification
+        pushNotificationSwitch.isChecked = presentation.getPushNotificationState()
```
check 에 대한 상태를 최초에만 가져오고,
그 이후에는 view 에 반영이 안되는 것 아닌가요~?

vagabond95 replied:
view 에 대한 상태변화는 presenter 로 부터 가져오는 것이 아니라,
presenter 가 '명령하는' 형태가 되어야합니다.

ippnsj replied:
switch 가 사용자에 의해 눌리면 view에 대한 상태변화(switch가 활성화됨)는 자동으로 이루어져서 presenter 에서 뷰에게 명령할 필요가 없다고 생각했습니다!
사용자의 입력에 의해 뷰의 상태가 이미 변했어도 presenter에서 뷰에 명령해야 하는 건가요?? 🤔

vagabond95 commented on app/src/main/java/woowacourse/movie/ui/fragment/movielist/adapter/TheaterAdapter.kt:
```diff
@@ -0,0 +1,27 @@
+    private val onTheaterItemViewClick: (Int) -> Unit = { onTheaterItemClick(theaters[it]) }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
+        val view = LayoutInflater.from(parent.context).inflate(R.layout.theater_item, parent, false)
```
데이터 바인딩 클래스를 이용하여 바로 inflate 를 해줘도 괜찮을 것 같습니다~!

[Ref](https://developer.android.com/topic/libraries/data-binding/generated-binding?hl=ko)

vagabond95 commented on app/src/main/java/woowacourse/movie/ui/activity/seatpicker/view/SeatPickerActivity.kt:
```diff
@@ -150,49 +101,70 @@ class SeatPickerActivity : AppCompatActivity() {
+        view: TextView,
+        seat: SeatModel
+    ) {
+        if (view.isSelected) {
+            presenter.cancelSeat(seat)
+            return
+        }
+        presenter.reserveSeat(seat)
```
cancel 할지, reserve 할 지 여부도 presenter 에서 판단할 수 는 없을까요?

vagabond95 replied:
이러한 비즈니스 로직을 view 에서 결정하게 되면,
presenter 가 비즈니스 로직을 중계하는 것이 아니라
view 의 로직을 단순히 옮긴 형태로 변질되기 쉬울 것 같아요.

vagabond95 commented on app/src/main/java/woowacourse/movie/ui/activity/seatpicker/view/SeatPickerActivity.kt:
```diff
@@ -15,47 +15,49 @@ import androidx.core.view.children
 
+    override fun onDestroy() {
+        super.onDestroy()
+        db.close()
```
db 를 activity 에서 다뤄야할 필요가 있을까요?

vagabond95 replied:
MVP 로 전환한다는 것은,
activity 를 곧 view 로 인식한다는 것과 비슷하다고 볼 수 있는데요.

이전 미션들에서 view 가 알아야하는 정보들이 무엇이었을지 고민해보시면 좋겠습니다!

vagabond95 commented on app/src/main/java/woowacourse/movie/ui/activity/seatpicker/presenter/SeatPickerPresenter.kt:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    override fun addReservation(db: SQLiteDatabase) {
```
db 를 presenter 생성 시 부터 가지고 있는 것은 어떨까요?

vagabond95 commented on app/src/main/java/woowacourse/movie/ui/activity/seatpicker/presenter/SeatPickerPresenter.kt:
```diff
@@ -0,0 +1,76 @@
+        view.afterReservation(ticketModel)
+    }
+
+    override fun getTicketModelWithOriginalPrice(): MovieTicketModel {
```
함수 네이밍에 구체적인 구현 내용이 들어갈 경우, 구현 내용이 바뀔 때마다 함수 이름도 바뀌어야합니다.

함수 네이밍은 '해야하는 일'만 표현해보는 것은 어떨까요?

vagabond95 commented on app/src/main/java/woowacourse/movie/ui/fragment/movielist/TheaterSelectorFragment.kt:
```diff
@@ -0,0 +1,61 @@
+        val theatersView = view.findViewById<RecyclerView>(R.id.rv_theater)
+        theatersView.adapter = TheaterAdapter(
+            movie.theaters,
+            onTheaterItemClick = { moveToDetailActivity(it) }
```
> 액티비티가 하는 일이 없어진다고 생각하여 굳이 Presenter 를 거칠 필요가 없다고 판단하였습니다

적어주신 부분이 이러한 부분 맞을까요? ㅎㅎ

아래 질문에 대해 답변해주시면, 이후에 이어서 논의를 이어나가도 좋겠습니다!

1. 액티비티가 하는일이 없어지는 것이 문제가 되는 일일까요?
2. (MVP) 패턴에서 액티비티 혹은 프래그먼트가 하는 일은 무엇일까요?
3. onTheaterItemClick 에서 `moveToDetailActivity()` 를 실행하기 전에 처리해줘야 하는 로직들이 늘어난다면 어떻게 구현하실  계획 이실까요?

ippnsj replied:
말씀해주신 부분도 비슷한 이유로 presenter 로 빼지 않았던 것 같아요!
제가 처음 이것에 대해 고민했던 상황은 MovieDetailActivity 에서 dataspinner의 값이 사용자에 의해 바뀌었을 때 바뀐 데이터를 model 에 저장할 필요가 없는 상황이었어요~!

리뷰어님의 질문에 대한 저의 답변은 다음과 같습니다!
1. 액티비티가 하는 일이 없어져서 문제이기 보다는 액티비티는 가벼워지는 대신 Presenter 의 역할이 방대해지는 문제가 있을 수 있다고 생각했습니다!
2. MVP 패턴에서 액티비티 혹은 프래그먼트의 역할은 View 에 해당한다고 생각합니다. 그래서 dataspinner 의 아이템이 바뀌는건 View 의 영역이라고 생각했어요!
3. 어떤 로직인지에 따라 다르겠지만 Model 의 값을 갱신해줘야 한다면 Presenter 에서 처리하고 그렇지 않다면 현재로서는 액티비티에서 처리하지 않을까 생각했어요!

vagabond95 replied:
https://github.com/woowacourse/android-movie-theater/pull/30#discussion_r1186907316

요 코멘트랑 같이 묶어서 이야기 나눠볼게요~!

우선, presenter 의 역할이 방대해질 수 있다는 점을 고민해주셔서 좋았습니다.

실제로 presenter 가 god - object 가 되는 일이 비일 비재하고, 문제점들이 나타나는 것이 사실입니다.
다만, 이는 presenter 를 다시 ui 컴포넌트 별로 쪼개거나, data (domain) layer 로 분리를 시키는 등 뒷단에서 해결할 일이지,
view 가 같이 부담할 일은 아니라고 생각합니다.

정리하면, view 의 역할과 presenter 은 역할은 명확하게 구분지어져 있고,  그것이 단순히 한쪽으로 쏠림을 우려하여 서로의 역할을 나눠갖는 것은 아키텍처로 레이어를 나누는 의미가 없다는 것이 제 생각입니다.

vagabond95 replied:
그렇다면 중요한 것은 view 의 역할과 presenter 의 역할을 정의하는 것이겠습니다.

사람마다 디테일은 다를 수 있지만, 규칙이 간단할 수록 적용하기 쉬우므로 저는 다음과 같이 생각하고있습니다.

- view 는 멍청합니다. 주어진 데이터를 표현하고, 입력을 전달합니다.
- 그 외에 모든 로직은 presenter 혹은 data layer 에서 처리할 일입니다.

vagabond95 replied:
말씀하신 2가지 케이스를 위 규칙을 토대로 살펴보겠습니다.

> switch 에서 check 여부를 presenter 에서 관리할 필요가 있나? 어차피 바로 ui 에 반영되지 않는가?

-> 현재의 구현에서는 타당한 구현 방식입니다. 그러나 이는 switch 의 on/off 여부를 view 가 스스로 결정하는 것이므로 위 규칙에 어긋나게 됩니다. 
다음과 같은 상황이 생기면 어떨까요?

switch 를 터치했을 때, 내부적으로 특정한 조건 a 를 만족할 때만 on / 그 외에는 error toast 를 띄운다.

현재의 구조에서는 스위치를 터치할때 곧 on/off 가 바뀌기 때문에 위 스펙을 구현할 수 없습니다.
반면, 위 규칙을 만족할 경우, view 입장에서는 1) check 조건이 바뀌었다. 2) check 여부는 presenter 에서 결정한다.
로 두게되면 쉽게 구현을 할 수 있습니다.

> 어떤 로직인지에 따라 다르겠지만 Model 의 값을 갱신해줘야 한다면 Presenter 에서 처리하고 그렇지 않다면 현재로서는 액티비티에서 처리하지 않을까 생각했어요!

클릭 처리를 하는 경우도 동일합니다. 
model 의 값을 갱신하는 처리가 아니더라도, 어떠한 선행조건이 걸린다면 현재 view 에서는 처리가 어렵습니다.
또한, "어떤 로직인지에 따라 다르겠지만" 이라는 전제 자체가 각 layer 의 역할이 모호하게 잡혀있다는 신호가 될 수 있습니다.
이럴 경우 layer 를 나누는 목적에서 벗어나 단순히 코드를 여러 파일로 분리한 이점밖에 얻지 못하게 됩니다.

vagabond95 replied:
요점은 이렇습니다.

view 의 상태는 현재도, 앞으로도 어떤 방식으로 컨트롤 될 지 알 수 없습니다.

이것을 view 에 위임한다고 당장 나쁜 구현이라고 이야기할수는 없습니다.  다만 presenter 라는 레이어를 두기로 했다면, 이것은 presenter 에서 관리하는 것이 추후에 변경 가능성을 고려했을 때, 가장 용이하다는 것이 MVP 혹은 MV-Whatever 아키텍처의 의의라고 저는 생각합니다.

vagabond95 replied:
글로에게 드리고 싶은 또 다른 얘기는, 

MVP 혹은 특정 아키텍처에 대해 너무 강한 선입견을 가지지 않으셨으면 합니다.

앞에서 이미 강하게 주장해놓고 무슨 소리이냐고 하실 수 있지만,
사실 위 내용은 제가 수년간 개발을 하면서 정립한 제 나름의 규칙입니다.

추후에 팀원들과 일하게 될 때 아키텍처에 대해 모두 통일된 의견을 가지면 좋겠으나,
중요하게 생각하는 부분이 모두 다르기 때문에 디테일한 부분까지 의견을 합일시키기가 어려운 것이 현실입니다.

따라서 '아~ 이렇게도 생각할 수 있겠구나' 정도로만 받아들여주시고,
글로가 추후에 개발을 계속하시면서 MVP 라는 아키텍처가 왜 나왔는지에 대해 이해하고,
글로만의 구현 디테일을 가져가되 팀원들과 핵심을 놓치지 않는 선에서 적절히 조율하시는 것이 가장 좋다고 생각합니다 :)

ippnsj replied:
리뷰어님~! 
MVP 에 대한 리뷰어님의 생각부터 당부의 말씀까지 너무너무 감사드립니다! 🤓
덕분에 MVP 를 어떻게 구현해야 할 지 감을 잡을 수 있었어요! (❁´◡`❁)
정말 감사해요~!!! 😊

vagabond95 commented:
양방향 바인딩은 데이터 바인딩의 기능 중 하나일 뿐, 사용하지 않는다고 해서 문제되는 부분은 없습니다.
현재는 적절한 수준으로 잘 활용해주셨다고 생각해요!

vagabond95 reviewed (changes requested):
적지 않은 변경사항이 있으셨을텐데 고생 많으셨습니다~!

질문 주신 내용을 포함하여 코멘트 남겼는데, 확인 부탁드릴게요 ㅎㅎ

vagabond95 reviewed (changes requested):
적지 않은 변경사항이 있으셨을텐데 고생 많으셨습니다~!

질문 주신 내용을 포함하여 코멘트 남겼는데, 확인 부탁드릴게요 ㅎㅎ

vagabond95 reviewed (changes requested):
적지 않은 변경사항이 있으셨을텐데 고생 많으셨습니다~!

질문 주신 내용을 포함하여 코멘트 남겼는데, 확인 부탁드릴게요 ㅎㅎ

vagabond95 commented:
> 어디까지를 Presenter의 역할로 두어야 할 지가 모호했습니다!

한번에 이해가 되기 어려운 개념이 맞습니다. 저도 일을하고 몇년간은 계속 이해하는 바가 달라졌으니까요 ㅎㅎ
https://github.com/woowacourse/android-movie-theater/pull/30#discussion_r1186815625
코멘트에 대해 답변 남겨주시면, 이후에 추가적으로 이야기 나눠봐요!

ippnsj committed:
refactor: TableRow 를 동적으로 추가하도록 수정

ippnsj committed:
refactor: MovieTicketActivity 에 DataBinding 추가

ippnsj committed:
refactor: 예매 내역을 가져오는 작업을 Application 레벨의 코드에서 분리

ippnsj committed:
refactor: presenter 가 view 를 참조하도록 수정

presenter 가 푸쉬 알림 상태를 체크하고 뷰에 업데이트를 명령하도록 수정

ippnsj committed:
refactor: SettingsFragment 에 DataBinding 적용

ippnsj committed:
refactor: DataBinding 을 하면서 동시에 inflate 하도록 수정

ippnsj committed:
refactor: 비즈니스 로직을 presenter 에서 수행하도록 수정

좌석 예약, 취소 여부를 presenter 에서 판단하도록 수정

ippnsj committed:
refactor: view 에서 database 를 다루지 않도록 수정

presenter 에 database 를 생성하여 넘긴 이후부터는 view 가 database 를 다루지 않도록 수정

ippnsj committed:
refactor: 함수명에서 구체적인 구현 내용 제거

ippnsj committed:
fix: 영화와 광고 개수 관련 버그 수정

MovieAdapter 의 아이템 개수를 영화 개수에 맞게 계산하도록 수정
광고 개수가 필요보다 적을 경우 default 광고를 보여주도록 수정

ippnsj committed:
refactor: SeatPickerPresenter 에서 database 의존성 제거

vagabond95 commented on app/src/main/java/woowacourse/movie/data/entity/Reservations.kt:
```diff
@@ -1,10 +1,16 @@
+import woowacourse.movie.data.reservation.ReservationRepository
 import woowacourse.movie.ui.model.MovieTicketModel
 
 object Reservations {
```
이전부터 궁금했던 부분인데, Reservations 은 일종의 캐시역할을 하는 클래스가 맞을까요?

캐시가 맞다면, repository 레벨에서 같이 관리되어도 좋지 않을까 라는 생각이 들었습니다.
캐시를 관리하는 레이어가 repository 를 주입받아 처리한다는 것이 조금 어색해서요!

ippnsj replied:
캐시역할을 하기 위해 둔 클래스가 맞습니다!
그래서 repository 를 주입받는 대신 MovieTicketModel 리스트를 받아서 restore 하도록 수정하였습니다~! 😊

vagabond95 commented on app/src/main/java/woowacourse/movie/ui/activity/seatpicker/view/SeatPickerActivity.kt:
```diff
@@ -0,0 +1,229 @@
+            tableRow.addView(seatLayout)
+        }
+        return tableRow
+    }
```
추후에는 요러한 구현을 수행하는 커스텀 뷰를 만들어봐도 좋겠습니다~!

vagabond95 commented on app/src/main/java/woowacourse/movie/ui/fragment/reservationlist/ReservationListFragment.kt:
```diff
@@ -41,9 +46,17 @@ class ReservationListFragment : Fragment() {
         setReservationView(reservationView)
     }
 
+    private fun setUpReservations() {
+        val db = DBHelper(requireContext()).readableDatabase
+        val dataSource = ReservationLocalDataSource(db)
+        val repository = ReservationRepository(dataSource)
+        Reservations.restore(repository)
+    }
```
presenter 를 두고, presenter 내부에서 요러한 처리를 해주는 것은 어떨까요?

vagabond95 commented on app/src/main/java/woowacourse/movie/ui/fragment/settings/view/SettingsFragment.kt:
```diff
@@ -0,0 +1,50 @@
+
+class SettingsFragment : Fragment(), SettingsContract.View {
+    override lateinit var presenter: SettingsContract.Presenter
+    private lateinit var binding: FragmentSettingsBinding
```
onDestroyView 생명주기 이후에 binidng 객체를 참조하면 무슨일이 일어날까요?

ippnsj replied:
![image](https://github.com/woowacourse/android-movie-theater/assets/49394114/d20f638a-53f7-41cd-ab81-bc5c560c27c9)
![image](https://github.com/woowacourse/android-movie-theater/assets/49394114/36e8f0b6-b4e0-44a9-a807-f099f8197f36)
잘 참조되는 것을 보니 view 가 제대로 삭제되지 못하고 있는 걸까요...?

vagabond95 replied:
슬랙 코멘트로 답변드렸어요!

vagabond95 reviewed (changes requested):
빠르게 코멘트 반영해주셨네요~!

MVP 에 대한 제 생각과 추가 코멘트 남겼어요. 확인 부탁드릴게요~!

ippnsj committed:
refactor: ReservationListFragment 에 MVP 적용

ippnsj committed:
refactor: Reservations 가 repository 대신 MovieTicketModel 리스트를 받도록 수정

ippnsj committed:
feat: SeatView 커스텀 뷰 생성 및 적용

vagabond95 reviewed (approved):
3, 4 단계에서 주어진 요구사항이 많아 쉽지 않으셨을텐데, 정말 고생 많으셨습니다~!

MVP 패턴을 이해하는데, 조금이라도 도움이 되셨길 바랍니다 ㅎㅎ

현재 미션은 종료하겠습니다. 남은 기간도 화이팅입니다~!

vagabond95 reviewed (approved):
3, 4 단계에서 주어진 요구사항이 많아 쉽지 않으셨을텐데, 정말 고생 많으셨습니다~!

MVP 패턴을 이해하는데, 조금이라도 도움이 되셨길 바랍니다 ㅎㅎ

현재 미션은 종료하겠습니다. 남은 기간도 화이팅입니다~!
